### DISCUSSION TITLE
ERROR: R8: java.lang.OutOfMemoryError: Java heap space

### DISCUSSION THREAD

**Discussion by damianzdunekinertia** [codemagic-ci-cd/codemagic-docs#2632] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

1. Upgrade to the latest version (8.6) of Gradle in the `gradle-wrapper.properties` file and the Android Gradle plugin (8.4.0) in the `android/build.gradle` file as advised on [that](https://docs.codemagic.io/troubleshooting/common-android-issues/#java-heap-space-out-of-memory-error-for-m1-builds) page.
2. Additionally upgrade to the latest Flutter SDK (3.22.2)
3. Try various configuration increasing Java heap space in the `gradle.properties`, like

```
org.gradle.jvmargs=-Xmx4096m
```
 
or

```
org.gradle.daemon=true
org.gradle.parallel=true
org.gradle.configureondemand=true
org.gradle.jvmargs=-Xmx4096m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```

### Expected results

Android build passes on M1 machine as I can't change it when using free plan.

### Actual results

Android build fails with error message:

```
java.lang.OutOfMemoryError: Java heap space
ERROR: R8: java.lang.OutOfMemoryError: Java heap space

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:minifyReleaseWithR8'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.R8Task$R8Runnable
   > Compilation failed to complete
```

### Build id (optional)

668fd087612f74af828bfd11

**Comment by himesh-cm** [comment]
[](url)Hey @damianzdunekinertia,
Can you try by modifying your android/gradlew file by adding this command? Let me know, how it goes.﻿​

```​exec "$JAVACMD" "${JVM_OPTS[@]}" -classpath "$CLASSPATH" org.gradle.wrapper.GradleWrapperMain -Dorg.gradle.jvmargs="-Xmx4096m" "$@"```
![image](https://github.com/user-attachments/assets/36f57fa2-be9b-43e1-8973-8e0e224349bc)

**Reply by damianzdunekinertia** [reply]
Hi @himesh-cm, thank you for your answer. 

Unfortunately, updating gradlew file doesn't help. It produces `
Unknown command-line option '-X'.` error (both on my local machine and CodeMagic). The latest build number is `669101f8a9867518374b0229`

I tried the different solutions (like below) similar to yours but it didn't work as well – at least on my local machine.

```
DEFAULT_JVM_OPTS="-Xmx4096m"

eval "set -- $(
    printf '%s\n' "$DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS" |
    xargs -n1 |
    sed 's~[^-[:alnum:]+,./:=@_]~\\&~g;' |
    tr '\n' ' '
)" '"$@"'

exec "$JAVACMD" "$@" -classpath "$CLASSPATH" org.gradle.wrapper.GradleWrapperMain
```

I'm attaching my [gradlew](https://github.com/user-attachments/files/16192354/gradlew.txt) file.

**Reply by damianzdunekinertia** [reply]
@himesh-cm do you have any idea why I've got `Unknown command-line option '-X'` error?

I'd be grateful for any suggestion how to make my build work.

**Reply by himesh-cm** [reply, marked as answer]
Hey @damianzdunekinertia 

After checking your gradlew file, please add the below snippet

```
set -- "-Dorg.gradle.jvmargs=-Xmx4096m" "$@"

exec "$JAVACMD" "$@" 
```
instead of 
`exec "$JAVACMD" "${JVM_OPTS[@]}" -classpath "$CLASSPATH" org.gradle.wrapper.GradleWrapperMain -Dorg.gradle.jvmargs="-Xmx4096m" "$@"`

Please try it locally first and let me know, if you run into any issues.

**Reply by damianzdunekinertia** [reply]
Thank you @himesh-cm, it worked!

**Comment by himesh-cm** [comment]
Super, While I’m closing our discussion for now, if you have any other questions or suggestions, feel free to reply back or start a new discussion anytime. Have a nice day!

**Comment by Neelansh-ns** [comment]
Different gradle.properties take priority.
More here: https://docs.gradle.org/current/userguide/build_environment.html#:~:text=JAVA_HOME-,Priority%20for%20configurations,-When%20configuring%20Gradle

So modifying the ```$HOME/.gradle``` gradle.properties should do the trick.

 Or ```GRADLE_OPTS``` can come in handy.
 
 
`GRADLE_OPTS="-Dorg.gradle.jvmargs=$GRADLE_JVMARGS $GRADLE_OPTS"`
`echo "export GRADLE_OPTS=\"$GRADLE_OPTS\"" >> $HOME/.bash_profile`

**Reply by hnvn** [reply]
I got this issue too. I tried to edit `gradlew` file and setup `GRADLE_OPTS`, nothing works for me

**Comment by hnvn** [comment]
I got it worked out finally by creating `gradle.properties` file inside `$HOME/.gradle`. BTW, editing `gradlew` not working for me because I modified it locally and this file is not uploaded to my git repo

**Reply by ChauCM** [reply]
What does your script look like when creating gradle.properties? After making one, my Java version changed from 17 to 11

Edited: Script for those needed


```
    - name: update gradle.properties in $HOME/.gradle
      script: |
        # Create or overwrite gradle.properties with new settings
        echo "org.gradle.jvmargs=-Xmx4096m" > $HOME/.gradle/gradle.properties
        echo "org.gradle.java.home=/Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home" >> $HOME/.gradle/gradle.properties

        # Show updated gradle.properties
        echo "Updated gradle.properties:"
        cat $HOME/.gradle/gradle.properties

```
